release-notes/10.0/preview/preview2/runtime.md

```diff
@@ -8,6 +8,167 @@
+|------------------------------------------------------------- |-----------:|------:|----------:|
+| foreach_static_readonly_array                                |   153.4 ns |  1.00 |         - |
+| foreach_opaque_array_via_interface                           |   843.2 ns |  5.50 |      32 B |
+| foreach_opaque_array_via_interface (no PGO)                  | 2,076.4 ns | 13.54 |      32 B |
+```
+
+Notice how `foreach_opaque_array_via_interface` allocates memory on the heap, suggesting the JIT failed to stack-allocate and promote the enumerator to registers. This is because the JIT relies on a technique called escape analysis to enable stack allocation. Escape analysis determines if an object's lifetime can exceed that of its creation context; if the JIT can guarantee an object will not outlive the current method, it can safely allocate it on the stack. In the above example, calling an interface method on the enumerator to control iteration causes it to escape, as the call takes a reference to the enumerator object. On the fast path of the type test, the JIT can try to devirtualize and inline these interface calls to keep the enumerator from escaping. However, escape analysis typically considers the whole method context, so the slow path's reliance on interface calls prevents the JIT from stack-allocating the enumerator at all.
+
+[dotnet/runtime #111473](https://github.com/dotnet/runtime/pull/111473) introduces conditional escape analysis -- a flow-sensitive form of the technique -- to the JIT. Conditional escape analysis can determine if an object will escape only on certain paths through the method, and prompt the JIT to create a fast path where the object never escapes. For array enumeration scenarios, conditional escape analysis reveals the enumerator will escape only when type tests for the collection fail, enabling the JIT to create a copy of the iteration code where the enumerator is stack-allocated and promoted. Once again, this reduces the abstraction cost considerably:
+```
+| Method                                                       | Mean       | Ratio | Allocated |
+|------------------------------------------------------------- |-----------:|------:|----------:|
+| foreach_static_readonly_array (.NET 9)                       |   150.8 ns |  1.00 |         - |
+| foreach_opaque_array_via_interface (.NET 9)                  |   874.7 ns |  5.80 |      32 B |
+| foreach_opaque_array_via_interface (.NET 10)                 |   277.9 ns |  1.84 |      32 B |
+```
+
+## Inlining of Late Devirtualized Methods
+
+The JIT compiler can replace virtual method calls with non-virtual equivalents when it can determine the exact type of the `this` object. However, this type information may not be available to the JIT unless a specific method call is inlined. Consider the following example:
+
+```cs
+IC obj = GetObject();
+obj.M();
+
+IC GetObject() => new C();
```

Comment:
If I understand correctly, it is usually devirt that enables inlining. In this case, it is the opposite, that we need inlining to enable devirt, which may allow further inlining. Fair?

Reply:
You're correct, the two tend to create virtuous cycles out of each other.